File: app/src/main/java/com/hk/ijournal/viewmodels/AccessViewModel.java
package com.hk.ijournal.viewmodels;

import android.app.Application;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.hk.ijournal.repository.AccessRepository;
import com.hk.ijournal.repository.models.AccessModel;
import com.hk.ijournal.repository.models.DiaryUser;

import java.time.LocalDate;

public class AccessViewModel extends AndroidViewModel implements AccessRepository.AccessDataResponse {
    //username and passcode are two way binding variables
    //login livedata
    public MutableLiveData<String> loginUsernameLive;
    public MutableLiveData<String> loginPasscodeLive;
    public LiveData<AccessModel.AccessValidation> loginUserValidation;
    public LiveData<AccessModel.AccessValidation> loginPasscodeValidation;
    //register livedata
    public MutableLiveData<String> registerUsernameLive;
    public MutableLiveData<String> registerPasscodeLive;
    public MutableLiveData<LocalDate> dobLiveData;
    public LiveData<AccessModel.AccessValidation> registerUserValidation;
    public LiveData<AccessModel.AccessValidation> registerPasscodeValidation;
    AccessRepository accessRepository;
    public MutableLiveData<AccessModel.AccessStatus> accessStatus;

    public AccessViewModel(@NonNull Application application) {
        super(application);
        Log.d("lifecycle", "accessVM constructor");
        accessRepository = new AccessRepository(application);
        bindLiveData();
    }

    private void bindLiveData() {
        //login binding
        loginUsernameLive = accessRepository.getLoginUsernameLive();
        loginPasscodeLive = accessRepository.getLoginPasscodeLive();
        loginUserValidation = accessRepository.getLoginUserValidation();
        loginPasscodeValidation = accessRepository.getLoginPasscodeValidation();
        //register binding
        registerUsernameLive = accessRepository.getRegisterUsernameLive();
        registerPasscodeLive = accessRepository.getRegisterPasscodeLive();
        registerUserValidation = accessRepository.getRegisterUserValidation();
        registerPasscodeValidation = accessRepository.getRegisterPasscodeValidation();
        dobLiveData = accessRepository.getDobLiveData();
    }

    public LiveData<AccessModel.AccessStatus> getAccessStatus() {
        if (accessStatus == null)
            accessStatus = new MutableLiveData<>();
        return accessStatus;
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public void onDateSelected(int year, int month, int dayOfMonth) {
        accessRepository.setDobLiveData(LocalDate.of(year, month, dayOfMonth));
    }

    public void loginUser() {
        accessRepository.loginUserAndSendAccessData(this);
    }

    public void registerUser() {
        accessRepository.registerUserAndSendAccessData(this);
    }

    @Override
    public void onAccessDataReceived(DiaryUser dbUser) {
        accessStatus.setValue(accessRepository.processAccessAndGetAccessStatus(dbUser));
    }
}

File: app/src/main/java/com/hk/ijournal/viewmodels/AccessViewModel.kt
package com.hk.ijournal.viewmodels

import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.lifecycle.*
import com.hk.ijournal.common.CommonLib.LOGTAG
import com.hk.ijournal.repository.AccessRepository
import com.hk.ijournal.repository.AccessRepositoryImpl
import com.hk.ijournal.repository.data.source.local.entities.User
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import java.time.LocalDate
import javax.inject.Inject

@HiltViewModel
class AccessViewModel @Inject constructor(private val accessRepositoryImpl: AccessRepository) : ViewModel() {

    //login livedata
    val loginUsernameLive: MutableLiveData<String> = MutableLiveData()
    val loginPasscodeLive: MutableLiveData<String> = MutableLiveData()

    //register livedata
    val registerUsernameLive: MutableLiveData<String> = MutableLiveData()
    val registerPasscodeLive: MutableLiveData<String> = MutableLiveData()
    val dobLiveData: MutableLiveData<LocalDate> = MutableLiveData()

    //access livedata
    val loginStatus: MutableLiveData<AccessRepositoryImpl.AccessUser> = MutableLiveData()
    val registerStatus: MutableLiveData<AccessRepositoryImpl.AccessUser> = MutableLiveData()

    fun getLoginUserValidation(): LiveData<AccessRepositoryImpl.AccessValidation> {
        //transforming based on username live data
        return Transformations.map(loginUsernameLive) {
            return@map if (accessRepositoryImpl.isUsernameInvalid(it)) AccessRepositoryImpl.AccessValidation.USERNAME_INVALID
            else
                null
        }
    }

    fun getLoginPasscodeValidation(): LiveData<AccessRepositoryImpl.AccessValidation> {
        //transforming based on passcode live data
        return Transformations.map(loginPasscodeLive) {
            return@map if (accessRepositoryImpl.isPassCodeInvalid(it)) AccessRepositoryImpl.AccessValidation.PASSCODE_INVALID
            else
                null
        }
    }

    fun getRegisterUserValidation(): LiveData<AccessRepositoryImpl.AccessValidation> {
        //transforming based on username live data
        return Transformations.map(registerUsernameLive) {
            return@map if (accessRepositoryImpl.isUsernameInvalid(it)) AccessRepositoryImpl.AccessValidation.USERNAME_INVALID
            else
                null
        }
    }

    fun getRegisterPasscodeValidation(): LiveData<AccessRepositoryImpl.AccessValidation> {
        //transforming based on passcode live data
        return Transformations.map(registerPasscodeLive) {
            return@map if (accessRepositoryImpl.isPassCodeInvalid(it)) AccessRepositoryImpl.AccessValidation.PASSCODE_INVALID
            else
                null
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    fun onDateSelected(year: Int, month: Int, dayOfMonth: Int) {
        dobLiveData.value = LocalDate.of(year, month + 1, dayOfMonth)
    }

    fun loginUser() = viewModelScope.launch {
        val diaryUser = User(loginUsernameLive.value.toString(), loginPasscodeLive.value.toString())
        val dbUser = accessRepositoryImpl.getMatchingUserinDb(loginUsernameLive.value.toString())
        dbUser.let {
            if (it.isSuccess) {
                loginStatus.value = accessRepositoryImpl.processLoginAndGetAccessStatus(it.getOrNull(), diaryUser)
            }
            else
                loginStatus.value = accessRepositoryImpl.processLoginAndGetAccessStatus(null, diaryUser)
        }
    }

    fun registerUser() = viewModelScope.launch {
        val diaryUser = User(registerUsernameLive.value.toString(), registerPasscodeLive.value.toString(), dobLiveData.value)
        val dbUser = accessRepositoryImpl.getMatchingUserinDb(registerUsernameLive.value.toString())
        dbUser.let {
            if (it.isSuccess) {
                registerStatus.value = accessRepositoryImpl.processRegisterAndGetAccessStatus(it.getOrNull(), diaryUser)
            } else
                registerStatus.value = accessRepositoryImpl.processRegisterAndGetAccessStatus(null, diaryUser)
        }
    }

    override fun onCleared() {
        super.onCleared()
        Log.d(LOGTAG, "AccessVM Cleared")
    }
}